Support for new variant IOT2050-SM

BaochengSu opened the pull request:
Introduction of new IOT2050-SM variant.

The main use case is to connect to some of the PLC1200 signal modules.

This variant drops the Arduino interface. A new ASIC is added to handle the communication between the AM6548 and the signal modules.

huaqianli committed:
mraa: Add support for the new IOT2050-SM variant

For mraa, the main difference between the new variant and the rest
is the tailoring down of the Arduino interface.

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

AsuraZeng committed:
node-red-gpio: Dynamic add user button label

Not all platform has the user button supported, so dynamic
to add the user button label.

Signed-off-by: chao zeng <chao.zeng@siemens.com>

AsuraZeng committed:
node-red: Adapt with the new board iot2050 sm

To adapt our new type of board SM. As It doesn't have arduino
interfaces on it.
So we just need to keep the led and user button control.

In the mraa introduced one new board type SM, according to the
board type to distinguish the interfaces between previous board
and the new board.

Signed-off-by: chao zeng <chao.zeng@siemens.com>

huaqianli committed:
firmware-update-package: Add making FWU package option

Firmware update package is good to generate within building time,
so add an image type to make the FWU package.

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

huaqianli committed:
iot2050-firmware-update: Refactor the firmware update tool

At the beginning, a additional firmware was intended to be integrated
into the upgrade tarball package. And the original code was kind of
process-oriented code, many common features are not extracted, many
duplicate codes exist, and the code coupling is very high, which is very
difficult to expand and maintain. It was necessary to refactor the code for
better maintainability and extension.

Anyway, in the end the additional firmware is planed to integrated into its
associated subsystem for decoupling. Refactored code is preserved for
possible  future maintainability and extension. For example, if new firmware
needs to be added in the future, whether it is the existing firmware upgrade
method or other upgrade methods, the previous upgrade logic or upgrade
method can be more easily introduced and reused.

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

BaochengSu committed:
Add Extended IO manager

Provide service to manage the extended IO (mainly PLC1200 signal modules). Including:

- Config the modules, or retrieve the configuration
- Sync system time to extended IO controller
- Update the firmware of the controller and some modules - currently
  only controller firmware updating is supported.
- Read extended IO events

The extended IO controller is an ASIC, whose job is to forward the
configuration to signal modules and collect events from modules. The
firmware of the controller provides the IO read/write functionalities
through a FUSE filesystem /eio. The iot2050-eiod.service is managing
this FUSE file system.

This manager is exposing service to localhost via gRPC, the
iot2050-eiosd.service is managing this gRPC server.

To sync the system time to eio controller, the
iot2050-eio-time-syncing.service is doing the job.

iot2050-eio-fwu-monitor.service is used to monitor the current firmware
version of the eio controller, and raise alarm to console when the
firmware does not match. This is useful when the host application is
upgraded but the eio controller firmware is not updated accordingly, for
example, when trying another example image via a SDCard/USB stick.

To consume the service provided by this manager, an iot2050-eio cli is
provided to config/retrieve/update-firmware from the console. This cli
is purely a gRPC client to call the RPC procedure provided by the
iot2050-eiosd.service.

It's also possible to consume the service from WEB applications, we will
provide a demo web application for configing/retrieving the
signal modules.

The module configuration is written in YAML, which will be validated via
JSON schemas internally.

For this YAML file, please check the
config-template/sm-config-example.yaml for details. Please note the
`description` for each slot, currently when retrieving the
configuration, the description should be a multiline text, however, it's
not so easy to generate multile line text for yaml. This part is still
WIP.

The YAML configuration is internally converted to the binary format which the
extended IO controller accepts. We may use some binary schema tools such
as Kaitai Struct to rewrite this part of logic in the future.

Co-developed-by: Li Hua Qian <huaqian.li@siemens.com>
Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

huaqianli committed:
board-conf-tools: Refine the iot2050setup

Arduino IOs and GPIOs are not needed on IOT2050 SM Variant.

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

BaochengSu committed:
switchserial: Increase the CP2102N reset delay to 1s

Per observation, CP2102N reset takes more than 30 ms to finish, increase
the assertion to 1 second to improve the stability.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

huaqianli committed:
Add event record service

This service is to read the power up, power loss, eio, tilt, uncover,
and watchdog reset events, then writes them into syslog and makes them
be readable for journald.

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

BaochengSu committed:
Add iot2050 configration web application

Currently the only supported feature is to config the PLC1200 modules.
Web application is easier to use and less error-prone compared to the
YAML configuration, but less efficient for mass deployment.

This also provides a base when new requirement of web application comes.

For the sake of familiarity, react + material UI is used, however in the
future, siemens/ix should be used to get a more pretty and unified UI
style.

This web app should also be possible to pack for the Host PC or server
to be a standalone version, the use case is to manipulate the
configuration but not deploy/retrieve.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

jan-kiszka committed:
Add SM device tree to DTB_FILES

Fixes UKI booting via UEFI, thus the SWUpdate image variant.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>

huaqianli committed:
node-red-node-random: Add node-red-node-random

Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>

BaochengSu committed:
Update SEBoot to D01.04.01.02

Changelog:

- Add retries if flashing reading error occurs
- Add DDR type in board information, print DDR size in boot log
- Add board support for IOT2050-SM variant
- Add 4G DDR memory support
- Adjust firewall configuration according to TI's latest pattern
- Fix AH12 pinmux for IOT2050-SM
- Shutdown both R5 cores on hand-over

Signed-off-by: Su Baocheng <baocheng.su@siemens.com>

BaochengSu committed:
linux-iot2050: Add the support for the SM variant

This patch set is the kernel bsp for the new IOT2050-SM variant.

Refer to the dts patch for details of the new variant.

For kernel part, add driver for Everlight PM16D17 light sensor.

In addition, a minor cleanup of the R5 lockstep disabling.

Signed-off-by: Chao Zeng <chao.zeng@siemens.com>
Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

BaochengSu committed:
u-boot-iot2050: Add the support for the SM variant

The dts for the new SM variant is idential to the kernel.

Since the introduction of the 4G DDR on SM variant, the memory size is
now determined by the SEBoot and then passed to the u-boot via scratchpad
board info.

Also the fdt fixup logic is generalized.

BTW a minor coding style change for the iot2050/board.c is hitchhiked.

Signed-off-by: Chao Zeng <chao.zeng@siemens.com>
Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0081-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,405 @@
++		&cp2102n_reset_pin_default
++		&main_pcie_enable_pins_default
++		&asic_spi_mux_ctrl_pin
++	>;
```
Check https://github.com/torvalds/linux/commit/f722090aeb70ca67b5f5a37a39a26fca29664dce

BaochengSu replied:
Reworked.

jan-kiszka commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
++	maximum-speed = "high-speed";
++	/delete-property/ snps,dis-u1-entry-quirk;
++	/delete-property/ snps,dis-u2-entry-quirk;
++};
```
This file is 100% identical to the kernel file (didn't check myself, that's why I ask)

jan-kiszka commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced.dtb";
+ 		else if(board_is_m2())
+ 			fdtfile = "ti/k3-am6548-iot2050-advanced-m2.dtb";
++		else if(board_is_sm())
```
Style: `if (...` - actually already for the m2 test above.

BaochengSu replied:
ACK

jan-kiszka commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
++	if (board_is_sm())
++		set_pinvalue("gpio@601000_22", "P3V3_PCIE_CON_EN", 1);
++	else
++		set_pinvalue("gpio@601000_17", "P3V3_PCIE_CON_EN", 1);
```
The label name changes here as well for existing M.2 - harmless?

jan-kiszka replied:
And this code is now run for all boards, not just M.2 as so far.

BaochengSu replied:
The plan is to control the 3.3V power pin of M.2 or miniPCIE interface, for all variants, to workaround some power timing issues observed on some cards. (Hardware change will be introduced one by one.)

The label name change is harmless. Since there is word 'M2' in the old name, which does not fit the fact of other variants. In general, either miniPCIE or M.2, is kind of PCIE connection, so maybe 'PCIE_CON' is a more suitable name. - There is indeed other kind of interfaces in these two connections, such as USB2/3 ... Any suggestion for a better naming?

BaochengSu replied:
Separated.

jan-kiszka commented on recipes-app/node-red-gpio/files/0001-Dynamic-add-user-button-label.patch:
```diff
@@ -0,0 +1,35 @@
+Subject: [PATCH] Dynamic add user button label
+
+Not all platform has the user button supported, so dynamic
+to add the user button label.
```
Sorry, does not parse yet for me: How exactly is that determined if the button is there or not? And what is "dynamic" about this?

AsuraZeng replied:
Adaptative for different board type.
This is fix for:
https://github.com/node-red/node-red-nodes/blob/25d7d668ef1123d1182ece9e6c2ccd4ddc3d3c03/hardware/intel/mraa-gpio-din.html#L71.

We add the user button for the IOT2050, but the raspi or organe pi don't have the user button label before,
This should be the specific for the iot2050.   So its need to according to the iot2050 board type to dynamic add this user button node.

jan-kiszka replied:
So, we broke this in upstream? Then please enhance the commit message accordingly so that we also submit it.

jan-kiszka commented on recipes-app/iot2050-eio-manager/README.md:
```diff
@@ -0,0 +1,24 @@
+# IOT2050 Extended IO Subsystem
```
Commit message still says "Complete the commit message" :wink:

BaochengSu replied:
Now it should be a reviewable states.

BaochengSu commented on recipes-app/switchserialmode/files/src/serial/cp2102n.c:
```diff
@@ -242,7 +242,7 @@ static void cp2102n_hardware_reset(void)
     if (ADVANCED_BOARD_PG1 != get_board_type()) {
         gpio_set("CP2102N-RESET", 0);
-        usleep(30 * 1000);
+        sleep(1);
```
The original 30ms was suggested by the datasheet and FAE, however it still introduced issue in the reboot loop testing.

BaochengSu commented on recipes-app/iot2050-firmware-update/iot2050-firmware-update_0.5.bb:
```diff
@@ -12,17 +12,23 @@
+    ln -sf ../lib/iot2050/fwu/iot2050-firmware-update ${D}/usr/sbin/iot2050-firmware-update
 }
 
 do_deploy_deb:append() {
```
@huaqianli , could you please help to answer?

BaochengSu replied:
All the firmware update tool refacts are folded into one commit.

BaochengSu commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
+ 	u8 mac_addr[8][ARP_HLEN];
+ 	char seboot_version[40 + 1];
++	u8 padding[3];
++	u32 ddr_size_mb;
```
Indeed, acked and will add it into the commit message of the separate patch

BaochengSu replied:
Separated and comments added into the commit message.

BaochengSu commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
+ {
+ 	if (board_is_m2())
+-		m2_fdt_fixup(blob);
++		variants_fdt_fixup(blob);
```
Maybe not a good time to do it anyway: I was thinking of promoting this part of logic to be more common... not only for m.2... just in case we have more overlays variants - what we'd done for SM but dropped just days before.

BaochengSu replied:
Or maybe you are wondering why this fixup is not applied to SM? - We've dropped all the overlays for SM due to:
1. USB3 issue
2. Some card uses the PERST pin even it's not using PCIE signal (4g card using usb2 signal), so a no-high-speed overlay is also not usable for such cards.

However I think this fdt fixup logic should be common - just in case we have another variant in the future that need overlay fixup as well.

jan-kiszka replied:
As above: The rework bits with reasoning should become separate.

BaochengSu replied:
Separated.

BaochengSu commented on recipes-bsp/u-boot/files/0006-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,634 @@
+
+*. DDR size is added into the the scratchpad board info.
+*. The new variant uses an different mPCIE/M2 power control pin
+
```
ACK

AsuraZeng commented on recipes-app/node-red-gpio/files/0002-Adaptee-new-iot2050-SM-type.patch:
```diff
@@ -0,0 +1,128 @@
+
+Adaptee new board iot2050 sm, new board reduce arduino interfaces.
+Only keep the led and button control. So need to remove those node
+in the node-red-gpios according to the board type
```
For the SM variant,  mraa only support led and user button.  So needing keep LED and user button

AsuraZeng commented on recipes-app/node-red-gpio/files/0002-Adaptee-new-iot2050-SM-type.patch:
```diff
@@ -0,0 +1,128 @@
++        onpaletteadd : function() {
++            $.getJSON('mraa-gpio/'+this.id,function(data) {
++                if (data === 27) {
++                    $("#red-ui-palette-GPIO div[data-palette-type='mraa-gpio-dout']").hide();
```
it not easy to init the those nodes label accroding to board-type. this is a workaroud to add the nodes before then according to the board type to hide the nodes  label

AsuraZeng commented on recipes-app/node-red-gpio/files/0002-Adaptee-new-iot2050-SM-type.patch:
```diff
@@ -0,0 +1,128 @@
++        onpaletteadd : function() {
++            $.getJSON('mraa-gpio/'+this.id,function(data) {
++                if (data === 27) {
++                    $("#red-ui-palette-GPIO div[data-palette-type='mraa-gpio-pwm']").hide();
```
as the explained before,  it is not easy to init the label according to the board type.

BaochengSu committed:
README.md: Add how to generate git patch files

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

BaochengSu commented on recipes-kernel/linux/files/patches-6.1/0081-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,405 @@
++&mcu_r5fss0 {
++	/* lock-step mode not supported on this board */
++	ti,cluster-mode = <0>;
++};
```
Reworked to make this common in a separate patch.

BaochengSu committed:
iot2050-conf-webui: sm1238: Hide E0-E3 module version temporarily

When selecting Ex as module version, the data variant range is from 0x81
- 0x84 in TIA portal, which is not coverred by the module manual.

Currently clarification is under going, meanwhile let's hide these
module versions before any feedback is received.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

BaochengSu committed:
iot2050-eio-manager: sm1238: Hide E0-E3 module version temporarily

When selecting Ex as module version, the data variant range is from 0x81
- 0x84 in TIA portal, which is not coverred by the module manual.

Currently clarification is under going, meanwhile let's hide these
module versions before any feedback is received.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>

jan-kiszka commented on README.md:
````diff
@@ -228,3 +228,14 @@ command can be used:
+
+```shell
+git format-patch --full-index --no-numbered --zero-commit --no-signature ...
+```
````
We already have CONTRIBUTING.md. I thinks this important hint better fits there.

BaochengSu replied:
Fixed in #507

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0081-iio-light-Add-support-for-everlight-pmd16d17-sensor.patch:
```diff
@@ -0,0 +1,408 @@
++				     struct iio_chan_spec const *chan,
++				     long mask)
++{
++	return 0;
```
Is that correct? I'm not finding other IIO drivers doing that. I rather think you are supposed to return some `IIO_VAL_*` constant.

BaochengSu replied:
Fixed in #509, the write interface is removed, since we don't need it at this moment. Also checked some similar drivers, write interface is not a must provided one.

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0081-iio-light-Add-support-for-everlight-pmd16d17-sensor.patch:
```diff
@@ -0,0 +1,408 @@
++		op_mode_setting_val |= PITIME_0_POINT_4_MS & PS_ITIME_MASK;
++	}
++
++	/* TBD: what if ps-wtime is not in dtb, this should return null */
```
See above.

BaochengSu replied:
Removed, fixed in https://github.com/siemens/meta-iot2050/pull/509

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0081-iio-light-Add-support-for-everlight-pmd16d17-sensor.patch:
```diff
@@ -0,0 +1,408 @@
++	struct iio_dev *indio_dev;
++	int ret;
++
++	/*request iio dev memory*/
```
Not a very informative comment.

BaochengSu replied:
Removed, fixed in #509

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0082-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,400 @@
++              - siemens,iot2050-advanced-sm
+               - siemens,iot2050-basic
+               - siemens,iot2050-basic-pg2
+               - ti,am654-evm
```
This must be a separate patch for upstream.

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0082-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,400 @@
++		&cp2102n_reset_pin_default
++		&main_pcie_enable_pins_default
++		&asic_spi_mux_ctrl_pin
++	>;
```
Wrong style, should be
```
<&cp2102n_reset_pin_default>,
<&main_pcie_enable_pins_default>,
...
```

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0082-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,400 @@
++		&push_button_pins_default
++		&db9_com_mode_pins_default
++		&soc_asic_pins
++	>;
```
see above

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0001-Watchdog-Support-WDIOF_CARDRESET-on-TI-AM65x-platfor.patch:
```diff
@@ -13,7 +13,7 @@ Signed-off-by: Li Hua Qian <huaqian.li@siemens.com>
 
 diff --git a/arch/arm/dts/k3-am65-iot2050-common.dtsi b/arch/arm/dts/k3-am65-iot2050-common.dtsi
-index 65da226847f..b6135b849f1 100644
+index 65da226847f45b594e9c7c251b298fee5933406b..b6135b849f1ac22ad263550069600a067a763794 100644
```
Forgot to use some of our patch generation rules?

BaochengSu replied:
Fixed in #509

jan-kiszka commented on README.md:
````diff
@@ -228,3 +228,14 @@ command can be used:
+example for linux kernel or u-boot, please use below command to generate patch:
+
+```shell
+git format-patch --full-index --no-numbered --zero-commit --no-signature ...
````
This is missing ``--abbrev=12``.

BaochengSu replied:
Fixed in #507

jan-kiszka commented on recipes-kernel/linux/files/patches-6.1/0082-dts-iot2050-Support-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,400 @@
++	/delete-property/ pinctrl-47;
++	/delete-property/ pinctrl-48;
++	/delete-property/ pinctrl-49;
++	/delete-property/ pinctrl-50;
```
Quite a lot to delete, not only here. It may be better to factor out some iot2050-arduino.dtsi and include that only from the devices that support it.

BaochengSu replied:
If factoring out. I find it's hard to deal with `arduino_i2c_aio_switch_pins_default` and `arduino_io_oe_pins_default`... Any ideas?

jan-kiszka commented on recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch:
```diff
@@ -0,0 +1,109 @@
+ 
++ulong board_get_usable_ram_top(ulong total_size)
++{
++#ifdef CONFIG_PHYS_64BIT
```
Isn't that always true for the IOT2050?

BaochengSu replied:
We only use u-boot for a53, so this is always true for us. I was thinking to be compatible with the R5 building, however, we don't provide R5 dts at all, this seems a fake compatibility.

Acked and will removed.

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch:
```diff
@@ -0,0 +1,109 @@
+ {
+-	if (board_is_advanced())
++	struct iot2050_info *info = IOT2050_INFO_DATA;
++#ifdef CONFIG_PHYS_64BIT
```
Also here: when is that false for our devices?

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0010-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,447 @@
+Subject: [PATCH] dts: iot2050: Support new IOT2050-SM variant
+
+the dts file for IOT2050-SM variant is copied from kernel side without
+any change.
```
Maybe some words on the new variant here would be nice, similar to the kernel patch.

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0010-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,447 @@
+ arch/arm/dts/k3-am65-iot2050-boot-image.dtsi  |   2 +-
+ .../k3-am6548-iot2050-advanced-sm-u-boot.dtsi |   1 +
+ .../arm/dts/k3-am6548-iot2050-advanced-sm.dts | 346 ++++++++++++++++++
+ board/siemens/iot2050/board.c                 |  15 +-
```
You will also have to update `doc/board/siemens/iot2050.rst`, I suspect.

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0009-board-siemens-iot2050-Generalize-the-fdt-fixup.patch:
```diff
@@ -0,0 +1,95 @@
+Since now we only have M.2 variant consuming the overlay, it may not
+have immediate effect for other variant, however this makes the future
+variant more easier to apply fdt fixups.
+
```
Update `doc/board/siemens/iot2050.rst`?

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-bsp/u-boot/files/0010-dts-iot2050-Support-new-IOT2050-SM-variant.patch:
```diff
@@ -0,0 +1,447 @@
+ 
+ 		fit@380000 {
+-			fit,fdt-list-val = "k3-am6528-iot2050-basic-pg2", "k3-am6548-iot2050-advanced-pg2", "k3-am6548-iot2050-advanced-m2";
++			fit,fdt-list-val = "k3-am6528-iot2050-basic-pg2", "k3-am6548-iot2050-advanced-pg2", "k3-am6548-iot2050-advanced-m2", "k3-am6548-iot2050-advanced-sm";
```
Does the dtc allow us to wrap this rather long line?

BaochengSu replied:
Fixed in #509

jan-kiszka commented on recipes-app/node-red-gpio/files/0002-Adaptee-new-iot2050-SM-type.patch:
```diff
@@ -0,0 +1,128 @@
+Date: Thu, 4 May 2023 11:51:20 +0800
+Subject: [PATCH] Adaptee new iot2050 SM type
+
+Adaptee new board iot2050 sm, new board reduce arduino interfaces.
```
"Adaptee"?

BaochengSu replied:
Fixed in #510

BaochengSu commented:
This PR is too big to review and rework, hence I will split it into multiple PRs.

huaqianli commented on recipes-devtools/firmware-update-package/firmware-update-package.bb:
```diff
@@ -0,0 +1,26 @@
+    # Generate the firmware update package
+    sh ${WORKDIR}/iot2050-generate-fwu-tarball.sh ${WORKDIR} \
+		${DEPLOY_DIR_IMAGE} $(${ISAR_RELEASE_CMD})
+}
```
ACK and versioned.

About dependency,  we have `do_create_tarball[depends] += "u-boot-iot2050:do_deploy"`,  is there any other dependencies you were suggesting?

jan-kiszka replied:
I missed that line while mentally filtering for `DEPENDS` - this is fine.

BaochengSu commented on recipes-kernel/linux/files/patches-6.1/0081-iio-light-Add-support-for-everlight-pmd16d17-sensor.patch:
```diff
@@ -0,0 +1,408 @@
++		break;
++	}
++
++	/* TBD: what if ps-itime is not in dtb, this should return null */
```
Removed, fixed in #509

BaochengSu commented on recipes-bsp/u-boot/files/0008-board-siemens-iot2050-Pass-DDR-size-from-FSBL.patch:
```diff
@@ -0,0 +1,109 @@
+Instead, the DDR size is determined by the FSBL(SEBOOT) then passed to
+u-boot through the scratchpad info.
+
+The SEBoot version must be >= D/V01.04.0x.0x to support this change.
```
Fixed in #509
